File: src/eYcel/encrypt.py
"""
Excel encryption pipeline.

Workflow
--------
1. Open the workbook (data_only=False so formulas are readable).
2. First pass: collect ALL unique text values to build global substitution map.
3. For each sheet, extract formulas, transform ALL cells, reinject formulas
   with text literals updated.
4. Save the encrypted workbook and rules.yaml.

Default behavior:
- ALL text (headers, titles, data) → consistent substitution (same word = same fake word)
- Text inside formulas → also substituted
- Numbers → kept unchanged (AI needs real values for formulas)
- Dates → kept unchanged
- Formulas → structure preserved, text literals inside updated
"""

import random
import string
from pathlib import Path
from typing import Any, Dict, Optional, Tuple
from datetime import datetime, timezone

import openpyxl
from openpyxl.utils import get_column_letter


from .formula_handler import extract_formulas, clear_formula_cells
from .transformations import (
    transform_hash,
    transform_offset_date,
    transform_offset_number,
    transform_scale,
    transform_shuffle,
    transform_keep,
    transform_anonymize,
    build_global_text_map,
    substitute_text_in_formula,
)
from .yaml_handler import generate_rules, save_rules


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _random_salt(length: int = 16) -> str:
    """Generate a random alphanumeric salt string."""
    return "".join(random.choices(string.ascii_letters + string.digits, k=length))


def _random_factor() -> float:
    """Return a random scale factor in [0.1, 0.9) ∪ (1.1, 2.0]."""
    factor = random.uniform(0.1, 2.0)
    while 0.9 <= factor <= 1.1:
        factor = random.uniform(0.1, 2.0)
    return round(factor, 6)


def _random_offset() -> float:
    """Return a random numeric offset in [-1000, -1) ∪ (1, 1000]."""
    return round(random.uniform(100, 1000) * random.choice([-1, 1]), 4)


def _random_day_offset() -> int:
    """Return a random day offset between ±365 days."""
    return random.randint(-365, 365)


def auto_detect_transform(
    col_values: list, scramble_numbers: bool = False, scramble_dates: bool = False,
) -> str:
    """
    Recommend a transform type based on cell values.

    New defaults:
    - string → "substitute" (consistent readable replacement)
    - int/float → "keep" (unless scramble_numbers=True → "scale")
    - date → "keep" (unless scramble_dates=True → "offset")
    """
    from .column_analyzer import detect_cell_type

    type_counts: Dict[str, int] = {}
    for v in col_values:
        if v is None or v == "":
            continue
        if isinstance(v, str) and v.startswith("="):
            continue
        ctype = detect_cell_type(v)
        type_counts[ctype] = type_counts.get(ctype, 0) + 1

    if not type_counts:
        return "keep"

    dominant = max(type_counts, key=type_counts.get)

    if dominant == "date":
        return "offset" if scramble_dates else "keep"
    if dominant in ("int", "float"):
        return "scale" if scramble_numbers else "keep"
    if dominant in ("string", "categorical"):
        return "substitute"
    return "keep"


def generate_output_paths(input_path: str) -> Tuple[str, str]:
    """Generate default output paths based on the input file name."""
    p = Path(input_path)
    stem = p.stem
    parent = p.parent
    encrypted = str(parent / f"{stem}_encrypted.xlsx")
    rules = str(parent / f"{stem}_rules.yaml")
    return encrypted, rules


# ---------------------------------------------------------------------------
# Column-level transform dispatcher
# ---------------------------------------------------------------------------

def _transform_cell(value: Any, config: Dict[str, Any], global_text_map: Dict[str, str]) -> Any:
    """Apply the correct forward transform to a single cell value."""
    # Text cells are ALWAYS substituted via the global map, regardless of column transform.
    # This ensures headers, titles, and text in numeric columns all get hidden.
    if isinstance(value, str) and not value.startswith("="):
        if value in global_text_map:
            return global_text_map[value]
        return value

    t = config.get("transform", "keep")
    if t == "hash":
        return transform_hash(value, config["salt"])
    if t == "offset":
        if hasattr(value, "year"):          # date / datetime
            return transform_offset_date(value, config["offset_days"])
        return transform_offset_number(value, config["offset"])
    if t == "scale":
        return transform_scale(value, config["factor"])
    if t == "shuffle":
        return transform_shuffle(value, config["mapping"])
    if t == "anonymize":
        return transform_anonymize(value, config.get("col_type", "string"))
    return transform_keep(value)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def _load_workbook_any_format(path: Path):
    """Load .xlsx, .xls, or .csv into an openpyxl Workbook."""
    suffix = path.suffix.lower()

    if suffix == ".xls":
        # Convert .xls → openpyxl Workbook via xlrd
        import xlrd
        xls = xlrd.open_workbook(str(path))
        wb = openpyxl.Workbook()
        # Remove default sheet
        wb.remove(wb.active)
        for sheet_idx in range(xls.nsheets):
            xls_sheet = xls.sheet_by_index(sheet_idx)
            ws = wb.create_sheet(title=xls_sheet.name)
            for row in range(xls_sheet.nrows):
                for col in range(xls_sheet.ncols):
                    cell = xls_sheet.cell(row, col)
                    value = cell.value
                    # xlrd cell types: 0=empty, 1=text, 2=number, 3=date, 4=bool, 5=error, 6=blank
                    if cell.ctype == 3:  # date
                        try:
                            from datetime import datetime as _dt
                            date_tuple = xlrd.xldate_as_tuple(value, xls.datemode)
                            value = _dt(*date_tuple)
                        except Exception:
                            pass
                    elif cell.ctype == 4:  # bool
                        value = bool(value)
                    elif cell.ctype in (0, 6):  # empty/blank
                        value = None
                    ws.cell(row=row + 1, column=col + 1, value=value)
        return wb

    elif suffix == ".csv":
        import csv as csv_mod
        wb = openpyxl.Workbook()
        ws = wb.active
        ws.title = path.stem
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            reader = csv_mod.reader(f)
            for row_idx, row_data in enumerate(reader, start=1):
                for col_idx, value in enumerate(row_data, start=1):
                    # Auto-detect types
                    v = value.strip()
                    if v == "":
                        continue
                    # Try int
                    try:
                        ws.cell(row=row_idx, column=col_idx, value=int(v))
                        continue
                    except ValueError:
                        pass
                    # Try float
                    try:
                        ws.cell(row=row_idx, column=col_idx, value=float(v))
                        continue
                    except ValueError:
                        pass
                    # String (including formulas starting with =)
                    ws.cell(row=row_idx, column=col_idx, value=v)
        return wb

    else:
        # .xlsx (default)
        return openpyxl.load_workbook(path, data_only=False)


def encrypt_excel(
    input_path: str,
    output_path: Optional[str] = None,
    rules: Optional[Dict[str, Any]] = None,
    scramble_numbers: bool = False,
    scramble_dates: bool = False,
) -> str:
    """
    Encrypt a spreadsheet file (.xlsx, .xls, or .csv).

    Default behavior:
    - ALL text → consistent substitution (same word = same fake word everywhere)
    - Numbers → kept unchanged (set scramble_numbers=True to scale them)
    - Dates → kept unchanged (set scramble_dates=True to offset them)
    - Formulas → preserved, text literals inside formulas also substituted

    Supports .xlsx, .xls (via xlrd), and .csv files.

    Returns:
        Path to the generated rules YAML file.
    """
    src = Path(input_path)
    if not src.exists():
        raise FileNotFoundError(f"Input file not found: {input_path}")

    default_encrypted, default_rules = generate_output_paths(input_path)
    if output_path is None:
        output_path = default_encrypted
    rules_path = str(
        Path(output_path).parent /
        (Path(output_path).stem.replace("_encrypted", "") + "_rules.yaml")
    )

    # ── 1. Open workbook (supports .xlsx, .xls, .csv) ─────────────────────
    wb = _load_workbook_any_format(src)

    # ── 2. First pass: collect ALL unique text values for global map ─────────
    all_texts = []
    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        for row in ws.iter_rows():
            for cell in row:
                if cell.value is None or cell.value == "":
                    continue
                if isinstance(cell.value, str):
                    if cell.value.startswith("="):
                        # Extract text literals from formulas
                        for literal in _extract_formula_literals(cell.value):
                            if literal and literal not in all_texts:
                                all_texts.append(literal)
                    else:
                        if cell.value not in all_texts:
                            all_texts.append(cell.value)

    global_text_map = build_global_text_map(all_texts)

    column_configs: Dict[str, Dict[str, Any]] = {}

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]

        max_row = ws.max_row or 1
        max_col = ws.max_column or 1

        # ── 3. Build per-column transform configs ────────────────────────────
        sheet_configs: Dict[int, Dict[str, Any]] = {}

        for col_idx in range(1, max_col + 1):
            col_letter = get_column_letter(col_idx)
            col_key = f"{sheet_name}!{col_letter}"

            # Collect ALL values in this column (all rows)
            col_values = []
            for row in range(1, max_row + 1):
                cell = ws.cell(row=row, column=col_idx)
                if cell.value is not None and not (
                    isinstance(cell.value, str) and cell.value.startswith("=")
                ):
                    col_values.append(cell.value)

            if rules and col_key in rules:
                cfg = dict(rules[col_key])
            else:
                transform = auto_detect_transform(col_values, scramble_numbers, scramble_dates)
                cfg = {"transform": transform}

                if transform == "scale":
                    cfg["factor"] = _random_factor()
                elif transform == "offset":
                    cfg["offset_days"] = _random_day_offset()
                    cfg["offset"] = _random_offset()

            column_configs[col_key] = cfg
            sheet_configs[col_idx] = cfg

        # ── 4. Extract formulas ──────────────────────────────────────────────
        formula_map = extract_formulas(ws)
        clear_formula_cells(ws, formula_map)

        # ── 5. Transform ALL cells in ALL rows ───────────────────────────────
        for row in range(1, max_row + 1):
            for col_idx, cfg in sheet_configs.items():
                cell = ws.cell(row=row, column=col_idx)
                if cell.value is None:
                    continue
                try:
                    cell.value = _transform_cell(cell.value, cfg, global_text_map)
                except Exception:
                    pass

        # ── 6. Reinject formulas with text substitutions ─────────────────────
        for (row, col), formula in formula_map.items():
            updated = substitute_text_in_formula(formula, global_text_map)
            ws.cell(row=row, column=col).value = updated

    # ── 7. Save encrypted workbook ───────────────────────────────────────────
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    wb.save(output_path)

    # ── 8. Store global inverse text map in rules ────────────────────────────
    inverse_map = {v: k for k, v in global_text_map.items()}
    column_configs["__global_text_map"] = {
        "transform": "substitute",
        "mapping": inverse_map,
    }

    # ── 9. Generate and save rules ───────────────────────────────────────────
    metadata = {
        "original_filename": src.name,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "sheets": wb.sheetnames,
    }
    rules_dict = generate_rules(metadata, column_configs)
    save_rules(rules_dict, rules_path)

    return rules_path


def _extract_formula_literals(formula: str) -> list:
    """Extract all string literals (text between double quotes) from a formula."""
    literals = []
    i = 0
    chars = list(formula)
    while i < len(chars):
        if chars[i] == '"':
            i += 1
            start = i
            while i < len(chars) and chars[i] != '"':
                i += 1
            literal = "".join(chars[start:i])
            if literal:
                literals.append(literal)
            i += 1
        else:
            i += 1
    return literals

File: eYcel-wasm/src/encrypt.rs
use std::collections::HashMap;

use rand::rngs::SmallRng;
use rand::SeedableRng;

use crate::cell_value::CellValue;
use crate::errors::EYcelError;
use crate::formula_handler::{clear_formula_cells, extract_formulas};
use crate::transformations::*;
use crate::xlsx_io::{col_index_to_letter, read_file, write_xlsx, SheetData};
use crate::yaml_handler::{ColumnConfig, Metadata, Rules};

/// Options controlling encryption behavior.
#[derive(Debug, Clone)]
pub struct EncryptOptions {
    /// If true, numeric columns are scaled by a random factor (reversible).
    /// If false (default), numbers are kept unchanged so AI can work with them.
    pub scramble_numbers: bool,
    /// If true, dates are offset by random days (reversible).
    /// If false (default), dates are kept unchanged.
    pub scramble_dates: bool,
}

impl Default for EncryptOptions {
    fn default() -> Self {
        Self {
            scramble_numbers: false,
            scramble_dates: false,
        }
    }
}

/// Apply the correct forward transform to a single cell value.
fn transform_cell(
    value: CellValue,
    config: &ColumnConfig,
    global_text_map: &HashMap<String, String>,
    rng: &mut SmallRng,
) -> CellValue {
    // Text cells are ALWAYS substituted via the global map, regardless of column transform.
    // This ensures headers, titles, and text in numeric columns all get hidden.
    if let CellValue::String(ref s) = value {
        if let Some(fake) = global_text_map.get(s) {
            return CellValue::String(fake.clone());
        }
        return value;
    }

    match config.transform.as_str() {
        "offset" => {
            if value.has_year() {
                match value {
                    CellValue::Date(d) => {
                        CellValue::Date(transform_offset_date(d, config.offset_days.unwrap_or(0)))
                    }
                    CellValue::DateTime(dt) => CellValue::DateTime(transform_offset_datetime(
                        dt,
                        config.offset_days.unwrap_or(0),
                    )),
                    _ => value,
                }
            } else if let Some(f) = value.as_f64() {
                CellValue::Float(transform_offset_number(f, config.offset.unwrap_or(0.0)))
            } else {
                value
            }
        }
        "scale" => {
            if let Some(f) = value.as_f64() {
                match transform_scale(f, config.factor.unwrap_or(1.0)) {
                    Ok(v) => CellValue::Float(v),
                    Err(_) => value,
                }
            } else {
                value
            }
        }
        "shuffle" => {
            if let Some(mapping) = &config.mapping {
                CellValue::String(transform_shuffle(&value.to_string_repr(), mapping))
            } else {
                value
            }
        }
        "anonymize" => {
            let col_type = config.col_type.as_deref().unwrap_or("string");
            transform_anonymize(&value, col_type, rng)
        }
        _ => value, // "keep" or unknown
    }
}

/// Replace text literals inside a formula string using the global text map.
/// E.g. `=COUNTIF(A:A, "Sales")` → `=COUNTIF(A:A, "Bako")` if "Sales"→"Bako".
fn substitute_text_in_formula(formula: &str, text_map: &HashMap<String, String>) -> String {
    let mut result = String::with_capacity(formula.len());
    let chars: Vec<char> = formula.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        if chars[i] == '"' {
            // Found start of a string literal — extract it
            let start = i + 1;
            i += 1;
            while i < chars.len() && chars[i] != '"' {
                i += 1;
            }
            let end = i;
            let literal: String = chars[start..end].iter().collect();

            // Check if this literal is in our text map
            if let Some(replacement) = text_map.get(&literal) {
                result.push('"');
                result.push_str(replacement);
                result.push('"');
            } else {
                result.push('"');
                result.push_str(&literal);
                result.push('"');
            }

            if i < chars.len() {
                i += 1; // skip closing quote
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }

    result
}

/// Encrypt a spreadsheet file (xlsx, xls, or csv).
///
/// **Default behavior:**
/// - ALL text in ALL cells (including headers, titles) → consistent substitution
/// - Text inside formulas (e.g. "Sales" in `=COUNTIF(A:A,"Sales")`) → also substituted
/// - Numbers → kept unchanged (AI needs real numbers for formulas)
/// - Dates → kept unchanged
/// - Formulas → structure preserved, text literals inside them updated
///
/// Use `options` to optionally scramble numbers and dates too.
pub fn encrypt_excel(
    input_bytes: &[u8],
    filename: &str,
    custom_rules: Option<&Rules>,
) -> Result<(Vec<u8>, String), EYcelError> {
    encrypt_excel_with_options(input_bytes, filename, custom_rules, &EncryptOptions::default())
}

pub fn encrypt_excel_with_options(
    input_bytes: &[u8],
    filename: &str,
    custom_rules: Option<&Rules>,
    options: &EncryptOptions,
) -> Result<(Vec<u8>, String), EYcelError> {
    let mut rng = SmallRng::from_entropy();

    // 1. Read the file
    let mut sheets = read_file(input_bytes, filename)?;

    // 2. First pass: collect ALL unique text values across ALL cells in ALL sheets
    //    (including headers, titles, and text literals inside formulas)
    let global_text_map = build_global_map_from_sheets(&sheets);

    let mut all_column_configs: HashMap<String, ColumnConfig> = HashMap::new();
    let mut sheet_names: Vec<String> = Vec::new();

    for sheet in &mut sheets {
        sheet_names.push(sheet.name.clone());

        if sheet.grid.is_empty() {
            continue;
        }

        let num_cols = sheet.num_cols();

        // 3. Build per-column transform configs
        //    Analyze ALL data in the column (all rows) to detect type
        let mut col_configs: Vec<(usize, ColumnConfig)> = Vec::new();

        for col_idx in 0..num_cols {
            let col_letter = col_index_to_letter(col_idx);
            let col_key = format!("{}!{}", sheet.name, col_letter);

            // Collect all cells in this column (all rows)
            let col_data: Vec<CellValue> = sheet
                .grid
                .iter()
                .map(|row| row.get(col_idx).cloned().unwrap_or(CellValue::Empty))
                .collect();

            let cfg = if let Some(rules) = custom_rules {
                if let Some(c) = rules.columns.get(&col_key) {
                    c.clone()
                } else if let Some(c) = rules.columns.get(&col_letter) {
                    c.clone()
                } else {
                    auto_config_for_column(&col_data, options, &mut rng)
                }
            } else {
                auto_config_for_column(&col_data, options, &mut rng)
            };

            all_column_configs.insert(col_key, cfg.clone());
            col_configs.push((col_idx, cfg));
        }

        // 4. Extract formulas BEFORE any transformation
        let formula_map = extract_formulas(&sheet.grid);

        // 5. Clear formula cells so transforms don't touch them
        clear_formula_cells(&mut sheet.grid, &formula_map);

        // 6. Transform ALL cells in ALL rows (no row is special)
        for row_idx in 0..sheet.grid.len() {
            for (col_idx, cfg) in &col_configs {
                if *col_idx < sheet.grid[row_idx].len() {
                    let cell = &sheet.grid[row_idx][*col_idx];
                    if cell.is_empty() {
                        continue;
                    }
                    let transformed =
                        transform_cell(cell.clone(), cfg, &global_text_map, &mut rng);
                    sheet.grid[row_idx][*col_idx] = transformed;
                }
            }
        }

        // 7. Reinject formulas, but with text substitutions applied inside them
        for (&(row, col), formula) in &formula_map {
            if row < sheet.grid.len() && col < sheet.grid[row].len() {
                let updated_formula = substitute_text_in_formula(formula, &global_text_map);
                sheet.grid[row][col] = CellValue::Formula(updated_formula);
            }
        }
    }

    // 8. Write encrypted workbook
    let encrypted_bytes = write_xlsx(&sheets)?;

    // 9. Store the global text map in rules (inverse: fake→original) for decryption
    let inverse_map: HashMap<String, String> = global_text_map
        .iter()
        .map(|(orig, fake)| (fake.clone(), orig.clone()))
        .collect();
    all_column_configs.insert(
        "__global_text_map".to_string(),
        ColumnConfig {
            transform: "substitute".to_string(),
            salt: None,
            factor: None,
            offset: None,
            offset_days: None,
            mapping: Some(inverse_map),
            col_type: None,
        },
    );

    // 10. Generate rules YAML
    let timestamp = chrono::Utc::now().to_rfc3339();
    let rules = Rules {
        metadata: Metadata {
            original_filename: filename.to_string(),
            timestamp,
            sheets: sheet_names,
        },
        columns: all_column_configs,
    };

    let rules_yaml = crate::yaml_handler::rules_to_yaml(&rules)
        .map_err(|e| EYcelError::Encryption(e))?;

    Ok((encrypted_bytes, rules_yaml))
}

/// Collect ALL unique text values from ALL cells in ALL sheets.
/// Also extracts text literals from inside formulas.
fn build_global_map_from_sheets(sheets: &[SheetData]) -> HashMap<String, String> {
    let mut unique_texts: Vec<String> = Vec::new();

    for sheet in sheets {
        for row in &sheet.grid {
            for cell in row {
                match cell {
                    CellValue::String(s) if !s.is_empty() => {
                        if !unique_texts.contains(s) {
                            unique_texts.push(s.clone());
                        }
                    }
                    CellValue::Formula(f) => {
                        // Extract text literals from formulas (strings inside quotes)
                        for literal in extract_formula_literals(f) {
                            if !literal.is_empty() && !unique_texts.contains(&literal) {
                                unique_texts.push(literal);
                            }
                        }
                    }
                    _ => {}
                }
            }
        }
    }

    build_global_text_map(&unique_texts)
}

/// Extract all string literals (text between double quotes) from a formula.
fn extract_formula_literals(formula: &str) -> Vec<String> {
    let mut literals = Vec::new();
    let chars: Vec<char> = formula.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        if chars[i] == '"' {
            i += 1;
            let start = i;
            while i < chars.len() && chars[i] != '"' {
                i += 1;
            }
            let literal: String = chars[start..i].iter().collect();
            if !literal.is_empty() {
                literals.push(literal);
            }
            i += 1; // skip closing quote
        } else {
            i += 1;
        }
    }

    literals
}

/// Auto-detect the best transform for a column.
fn auto_config_for_column(
    col_data: &[CellValue],
    options: &EncryptOptions,
    rng: &mut SmallRng,
) -> ColumnConfig {
    // Detect dominant type
    let mut type_counts: HashMap<&str, usize> = HashMap::new();
    for cell in col_data {
        if !cell.is_empty() && !cell.is_formula() {
            *type_counts.entry(cell.type_name()).or_insert(0) += 1;
        }
    }

    let dominant = type_counts
        .iter()
        .max_by_key(|(_, count)| *count)
        .map(|(t, _)| *t)
        .unwrap_or("empty");

    match dominant {
        "string" => ColumnConfig {
            transform: "substitute".to_string(),
            salt: None,
            factor: None,
            offset: None,
            offset_days: None,
            mapping: None,
            col_type: None,
        },
        "int" | "float" => {
            if options.scramble_numbers {
                ColumnConfig {
                    transform: "scale".to_string(),
                    salt: None,
                    factor: Some(random_factor(rng)),
                    offset: None,
                    offset_days: None,
                    mapping: None,
                    col_type: None,
                }
            } else {
                ColumnConfig {
                    transform: "keep".to_string(),
                    salt: None,
                    factor: None,
                    offset: None,
                    offset_days: None,
                    mapping: None,
                    col_type: None,
                }
            }
        }
        "date" => {
            if options.scramble_dates {
                ColumnConfig {
                    transform: "offset".to_string(),
                    salt: None,
                    factor: None,
                    offset: Some(random_offset(rng)),
                    offset_days: Some(random_day_offset(rng)),
                    mapping: None,
                    col_type: None,
                }
            } else {
                ColumnConfig {
                    transform: "keep".to_string(),
                    salt: None,
                    factor: None,
                    offset: None,
                    offset_days: None,
                    mapping: None,
                    col_type: None,
                }
            }
        }
        _ => ColumnConfig {
            transform: "keep".to_string(),
            salt: None,
            factor: None,
            offset: None,
            offset_days: None,
            mapping: None,
            col_type: None,
        },
    }
}
